File: backend/api/exploration.py
"""Exploration API endpoints - Track user's learning journey"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional
from storage.exploration_store import exploration_store

router = APIRouter(prefix="/exploration", tags=["exploration"])


class RecordQueryRequest(BaseModel):
    """Request to record a query in exploration history"""
    notebook_id: str
    query: str
    topics: List[str] = []
    sources_used: List[str] = []
    confidence: float = 0.5
    answer_preview: str = ""


@router.post("/record")
async def record_query(request: RecordQueryRequest):
    """Record a query as part of the user's exploration journey"""
    result = await exploration_store.record_query(
        notebook_id=request.notebook_id,
        query=request.query,
        topics=request.topics,
        sources_used=request.sources_used,
        confidence=request.confidence,
        answer_preview=request.answer_preview
    )
    return {"status": "recorded", "query_id": result["id"]}


@router.get("/journey/{notebook_id}")
async def get_journey(notebook_id: str, limit: int = 50):
    """Get the exploration journey for a notebook"""
    return await exploration_store.get_journey(notebook_id, limit)


@router.get("/suggestions/{notebook_id}")
async def get_suggestions(notebook_id: str):
    """Get suggestions for continuing exploration"""
    return await exploration_store.get_suggestions(notebook_id)


@router.delete("/clear/{notebook_id}")
async def clear_exploration(notebook_id: str):
    """Clear exploration history for a notebook"""
    await exploration_store.clear_notebook(notebook_id)
    return {"status": "cleared"}

File: src/services/exploration.ts
/**
 * Exploration Service - Track and retrieve user's learning journey
 */
import { API_BASE_URL } from './api';

export interface QueryRecord {
    id: string;
    query: string;
    topics: string[];
    sources_used: string[];
    confidence: number;
    answer_preview: string;
    timestamp: string;
}

export interface TopicExplored {
    name: string;
    count: number;
    first_seen: string;
    last_seen: string;
}

export interface SourceAccessed {
    source_id: string;
    count: number;
    first_accessed: string;
    last_accessed: string;
}

export interface JourneyResponse {
    notebook_id: string;
    queries: QueryRecord[];
    topics_explored: TopicExplored[];
    sources_accessed: SourceAccessed[];
    total_queries: number;
}

export interface Suggestion {
    type: 'continue' | 'dive_deeper';
    message: string;
    query?: string;
    topic?: string;
}

export interface SuggestionsResponse {
    suggestions: Suggestion[];
    shallow_topics: { name: string; count: number }[];
}

export const explorationService = {
    /**
     * Record a query in the exploration history
     */
    async recordQuery(
        notebookId: string,
        query: string,
        topics: string[] = [],
        sourcesUsed: string[] = [],
        confidence: number = 0.5,
        answerPreview: string = ''
    ): Promise<{ status: string; query_id: string }> {
        const response = await fetch(`${API_BASE_URL}/exploration/record`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                notebook_id: notebookId,
                query,
                topics,
                sources_used: sourcesUsed,
                confidence,
                answer_preview: answerPreview,
            }),
        });

        if (!response.ok) {
            throw new Error('Failed to record query');
        }

        return response.json();
    },

    /**
     * Get the exploration journey for a notebook
     */
    async getJourney(notebookId: string, limit: number = 50): Promise<JourneyResponse> {
        const response = await fetch(`${API_BASE_URL}/exploration/journey/${notebookId}?limit=${limit}`);

        if (!response.ok) {
            throw new Error('Failed to get journey');
        }

        return response.json();
    },

    /**
     * Get suggestions for continuing exploration
     */
    async getSuggestions(notebookId: string): Promise<SuggestionsResponse> {
        const response = await fetch(`${API_BASE_URL}/exploration/suggestions/${notebookId}`);

        if (!response.ok) {
            throw new Error('Failed to get suggestions');
        }

        return response.json();
    },

    /**
     * Clear exploration history for a notebook
     */
    async clearHistory(notebookId: string): Promise<void> {
        const response = await fetch(`${API_BASE_URL}/exploration/clear/${notebookId}`, {
            method: 'DELETE',
        });

        if (!response.ok) {
            throw new Error('Failed to clear history');
        }
    },
};
